added "default muting" and "remember muting" option.

johann-gambol opened the pull request:
I added two more option to the general settings tab, addressing the media:

-  default muting: All videos will be muted by default

 - remember muting in feed: if the user mutes or enables the sound for one video in the feed, the decision will be remembered across the feed. 

Both options default to "false".
I find them very useful.

jonjomckay commented on lib/settings/_general.dart:
```diff
@@ -254,6 +254,18 @@ class SettingsGeneralFragment extends StatelessWidget with AppBarMixin {
+        ///TODO: translation etc.
+        const PrefSwitch(
+          pref: optionsMediaDefaultMute,
+          title: Text("Default muting"),
```
`Mute videos`

jonjomckay commented on lib/settings/_general.dart:
```diff
@@ -254,6 +254,18 @@ class SettingsGeneralFragment extends StatelessWidget with AppBarMixin {
+        const PrefSwitch(
+          pref: optionsMediaDefaultMute,
+          title: Text("Default muting"),
+          subtitle: Text("Mute media by default"),
```
`Whether all videos should be muted`

jonjomckay commented on lib/settings/_general.dart:
```diff
@@ -254,6 +254,18 @@ class SettingsGeneralFragment extends StatelessWidget with AppBarMixin {
+          subtitle: Text("Mute media by default"),
+        ),
+        const PrefSwitch(
+          pref: optionMediaRememberMute,
```
I don't think this need to be an option - it probably makes things a bit more confusing to the user. I think we should just automatically keep the user's current mute setting while a user is scrolling through a group feed.

johann-gambol replied:
I really like this option - in fact there is an reddit app/client, called "infinity", where this is an option. And i got used to it pretty fast. And if i'm not wrong, the twitter App itself behaves this way. But i'm totally okay with it, removing it.

jonjomckay commented on lib/tweet/_video.dart:
```diff
@@ -99,9 +103,34 @@ class _TweetVideoState extends State<TweetVideo> {
     );
   }
 
+  void _loadMuteOptions() {
```
I think the content of this method should be inlined in `_loadVideo` - mostly because it otherwise introduces a race, where the video might start playing before the "mute" code runs, meaning audio blares for a moment before stopping.

We should be able to create the controller as normal, but disable `autoPlay`, then set the volume on the controller instance, then start the video playing.

One other thing that will need to be done is to store the `shouldMuteFeed` value inside `_SubscriptionGroupFeedState`, which means passing it down until it gets into this widget - or creating a `Provider` around `_SubscriptionGroupFeedState`, storing the preference in a per-group instance store, which would mean we don't need to pass down through the entire context.

jonjomckay reviewed (changes requested):
Thank you for this! I've left a few comments for you.

Please let me know if they make sense or not, or if you'd like me to make the changes on your behalf 🙂

johann-gambol commented:
Thanks for the reply! I'll try to add the changes.

johann-gambol commented:
I applied those changes - if I got you right.

- I created a `VideoMuteModel` as a `StreamStore` and instantiated it in "main.dart"
- I introduced a `MuteSettings` data class
- the mute settings will be loaded in  `_SubscriptionGroupFeedState`'s `initState` function
- the mute settings will now be checked inside the _loadVideo function

For now I stayed with the "remember Mute" option. I thought about it and it still makes sense to me. Otherwise the user would need to enable the sound for every video in the feed, if he goes with the "default mute" option. And I can imagine it quite annoying, if the feed has a lot of Videos. Otherwise you would need to disable "default mute" option in the settings menu and afterwards enable it again. Or instead, the feed just remembers your decision for the session. Does this makes sense?
And because it's an option, that's turned off by default, the user can self use to give it a try.

jonjomckay commented:
Thanks @johann-gambol!

I think there might have been a bit of confusion with my request regarding the "remember mute" option. I'm not suggesting to remove the option and the code - only the option, essentially inlining "remember mute" as the normal behaviour. This is why I suggested injecting a "temporary" instance using a `Provider`, similar to how `TweetContextState` works, by storing the setting per-context, with a default of the global setting.

Having a preference isn't free, and I think it makes more sense to start without one, and we can add one later is there's demand for it, otherwise it makes things a bit harder to maintain.

johann-gambol commented:
Ok! I removed "remember muting" as an option and made it the default behavior. I'm not sure, what you mean with `TweetContextState`, but the `MuteSettings` are now implemented in a `VideoMuteModel`, which is distributed by a `Provider`. But i guess it still not implemented the way you meant it to be.
